agent/service_manager.go

```diff
@@ -18,100 +18,200 @@ import (
+		return s.agent.addServiceInternal(&addServiceRequest{
+			service:               service,
+			chkTypes:              chkTypes,
+			persistService:        nil,
+			persistDefaults:       nil,
+			persist:               persist,
+			persistServiceConfig:  false,
+			token:                 token,
+			replaceExistingChecks: replaceExistingChecks,
+			source:                source,
+		})
 	}
 
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	reg := serviceRegistration{
-		service:  service,
-		chkTypes: chkTypes,
-		persist:  persist,
-		token:    token,
-		source:   source,
+	reg := &serviceRegistration{
+		service:               service,
+		chkTypes:              chkTypes,
+		persist:               persist,
+		token:                 token,
+		replaceExistingChecks: replaceExistingChecks,
+		source:                source,
 	}
 
-	// If a service watch already exists, update the registration. Otherwise,
-	// start a new config watcher.
-	watch, ok := s.services[service.ID]
-	if ok {
-		if err := watch.updateRegistration(&reg); err != nil {
-			return err
-		}
-		s.agent.logger.Printf("[DEBUG] agent.manager: updated local registration for service %q", service.ID)
-	} else {
-		// This is a new entry, so get the existing global config and do the initial
-		// registration with the merged config.
-		watch := &serviceConfigWatch{
-			registration: &reg,
-			readyCh:      make(chan error),
-			updateCh:     make(chan cache.UpdateEvent, 1),
-			agent:        s.agent,
-		}
+	s.servicesLock.Lock()
+	defer s.servicesLock.Unlock()
 
-		// Start the config watch, which starts a blocking query for the resolved service config
-		// in the background.
-		if err := watch.Start(); err != nil {
-			return err
-		}
+	// If a service watch already exists, shut it down and replace it.
+	oldWatch, updating := s.services[service.ID]
+	if updating {
+		oldWatch.Stop()
+		delete(s.services, service.ID)
+	}
 
-		// Call ReadyWait to block until the cache has returned the initial config and the service
-		// has been registered.
-		if err := watch.ReadyWait(); err != nil {
-			watch.Stop()
-			return err
-		}
+	// Get the existing global config and do the initial registration with the
+	// merged config.
+	watch := &serviceConfigWatch{
+		registration: reg,
+		updateCh:     make(chan cache.UpdateEvent, 1),
+		agent:        s.agent,
+		registerCh:   s.registerCh,
+	}
 
-		s.services[service.ID] = watch
+	err := watch.RegisterAndStart(
+		previousDefaults,
+		waitForCentralConfig,
+		persistServiceConfig,
+		s.ctx,
+		&s.running,
+	)
+	if err != nil {
+		return err
+	}
+
+	s.services[service.ID] = watch
 
+	if updating {
+		s.agent.logger.Printf("[DEBUG] agent.manager: updated local registration for service %q", service.ID)
+	} else {
 		s.agent.logger.Printf("[DEBUG] agent.manager: added local registration for service %q", service.ID)
 	}
 
 	return nil
 }
 
+// NOTE: must hold the Agent.stateLock!
```

Comment:
This confuses me - the method holds the manager's `servicesLock`. Do you mean the caller must hold the state lock when calling?

Reply:
Yeah i meant that the caller should hold the lock.